src/controllers/mainController.ts

```diff
@@ -113,10 +115,38 @@ export default class MainController implements vscode.Disposable {
                 this.registerCommand(Constants.cmdShowGettingStarted);
                 this._event.on(Constants.cmdShowGettingStarted, () => { self.launchGettingStartedPage(); });
                 this.registerCommand(Constants.cmdNewQuery);
-                this._event.on(Constants.cmdNewQuery, () => { self.runAndLogErrors(self.onNewQuery(), 'onNewQuery'); });
+                this._event.on(Constants.cmdNewQuery, (sessionId?: string) => { self.runAndLogErrors(self.onNewQuery(sessionId), 'onNewQuery'); });
                 this.registerCommand(Constants.cmdRebuildIntelliSenseCache);
                 this._event.on(Constants.cmdRebuildIntelliSenseCache, () => { self.onRebuildIntelliSense(); });
 
+                // register the object explorer tree provider
+                this._objectExplorerProvider = new ObjectExplorerProvider(this._connectionMgr);
+                this._context.subscriptions.push(
+                    vscode.window.registerTreeDataProvider('objectExplorer', this._objectExplorerProvider)
+                );
+                this.registerCommand(Constants.cmdAddObjectExplorer);
+                this._event.on(Constants.cmdAddObjectExplorer, async () => {
+                    if (self._objectExplorerProvider.objectExplorerExists) {
+                        return self._objectExplorerProvider.createSession();
+                    } else {
+                        self._objectExplorerProvider.objectExplorerExists = true;
+                        return vscode.commands.executeCommand('workbench.view.extension.objectExplorer');
```

Comment:
This seems odd to have it set here. Shouldn't the objectExplorerProvider be the one setting this whenever it's ready?

Reply:
What does this even control anyways? I'm not sure exists is the right term here - it seems like this is more saying whether the provider is ready or not?

---
In reply to: [303634495](https://github.com/microsoft/vscode-mssql/pull/1259#discussion_r303634495) [](ancestors = 303634495)

Reply:
This is just for the case when a user opens the `Object Explorer` for the first time. The provider doesn't know that yet.